staging/src/k8s.io/component-base/logs/logrotation/logrotation.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logrotation
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+const timeLayout string = "20060102-150405"
+
+type RotationFile struct {
+	// required, the max size of the log file in MB, 0 means no rotation
+	maxSize uint
+	// required, the max age of the log file in days, 0 means no cleanup
+	maxAge      uint
+	filePath    string
+	mut         sync.Mutex
+	file        *os.File
+	currentSize int64
+	isFlushed   bool
+	stop        chan struct{}
+}
+
+func Open(filePath string, enableFlush bool, maxSize uint, maxAge uint) (w *RotationFile, err error) {
+	w = &RotationFile{
+		filePath: filePath,
+		maxSize:  maxSize,
+		maxAge:   maxAge,
+	}
+
+	logFile, err := os.OpenFile(w.filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+
+	w.file = logFile
+
+	w.isFlushed = true
+	if enableFlush {
+		interval := time.NewTicker(time.Second * 5)
+		w.stop = make(chan struct{})
+
+		go func() {
+			defer interval.Stop()
+			for {
+				select {
+				case <-interval.C:
+					// This block runs every time the ticker ticks
+					w.isFlushed = false
```

Comment:
Data race with `if !w.isFlushed` in `Write`.

Reply:
Do you think it is tolerable for some race condition here? As even it happens, the next tick will revert it back so we will guarantee the flushing, plus I would assume this race condition will be rare case?

Reply:
added the lock, it will have values for the locking. one of the scenarios is to avoid flaky UT tests

Reply:
Data races cannot be tolerated. They lead to undefined behavior and test failures - assuming that there are tests. If there are no tests yet which exercise this aspect, then they should get added.